---
title: "WormBase Biomart Information"
output: html_document
---

```{r setup}
library(biomaRt)
library(plyr)
library(seqcloudR)
```

```{r functions}
bm2wb <- function(df) {
  colnames(df) <- camel(colnames(df))
  rownames(df) <- df$ensemblGeneId
  source("geneIdRows.R", local = TRUE)
  df$ensemblGeneId <- NULL
  return(df)
}
```


## Connect to Biomart

* `entrezgene` = Entrez identifier
* `external_gene_name` = Ensembl public name
* `wormbase_locus` = WormBase public name

Use `wormbase_gene_seq_name` for clean sequence identifier.

```{r connect}
mart <- useMart("ensembl", "celegans_gene_ensembl")
#! biomartOptions <- listAttributes(mart)
```


## Simple gene length info

```{r basic}
df <- getBM(mart = mart,
            attributes = c("ensembl_gene_id",
                           "gene_biotype",
                           "chromosome_name",
                           "start_position",
                           "end_position",
                           "strand",
                           "description"))
basic <- bm2wb(df)
```


## Entrez identifiers

```{r entrez}
df <- getBM(mart = mart,
            attributes = c("ensembl_gene_id",
                           "entrezgene"))
df <- ddply(df, .(ensembl_gene_id), summarize,
            entrezgene = paste(sort(unique(entrezgene)), collapse = ", "))
entrezGeneId <- bm2wb(df)
```


## Refseq identifiers

```{r refseq}
df <- getBM(mart = mart,
            attributes = c("ensembl_gene_id",
                           "refseq_mrna",
                           "refseq_ncrna"))
df <- ddply(df, .(ensembl_gene_id), summarize,
            refseq_mrna = paste(sort(unique(refseq_mrna)), collapse = ", "),
            refseq_ncrna = paste(sort(unique(refseq_ncrna)), collapse = ", "))
refseq <- bm2wb(df)
```


## UniProt identifiers

```{r uniprot}
df <- getBM(mart = mart,
            attributes = c("ensembl_gene_id",
                           "uniprot_sptrembl",
                           "uniprot_swissprot"))
df <- ddply(df, .(ensembl_gene_id), summarize,
            uniprot_sptrembl = paste(sort(unique(uniprot_sptrembl)), collapse = ", "),
            uniprot_swissprot = paste(sort(unique(uniprot_swissprot)), collapse = ", "))
uniprot <- bm2wb(df)
```


## Homology

```{r homology}
df <- getBM(mart = mart,
            attributes = c("ensembl_gene_id",
                           "hsapiens_homolog_ensembl_gene"))
df <- ddply(df, .(ensembl_gene_id), summarize,
            hsapiens_homolog_ensembl_gene = paste(sort(unique(hsapiens_homolog_ensembl_gene)), collapse = ", "))
homology <- bm2wb(df)
```


## Gene Ontology

```{r ontology}
df <- getBM(mart = mart,
            attributes = c("ensembl_gene_id",
                           "go_id",
                           "name_1006"))
df <- ddply(df, .(ensembl_gene_id), summarize,
            go_id = paste(sort(unique(go_id)),collapse = ", "),
            name_1006 = paste(sort(unique(name_1006)), collapse = ", "))
geneOntology <- bm2wb(df)
```



## Interpro

```{r interpro}
df <- getBM(mart = mart,
            attributes = c("ensembl_gene_id",
                           "interpro",
                           "interpro_short_description",
                           "interpro_description"))
df <- ddply(df, .(ensembl_gene_id), summarize,
            interpro = paste(sort(unique(interpro)),collapse = ", "),
            interpro_short_description = paste(sort(unique(interpro_short_description)), collapse = ", "),
            interpro_description = paste(sort(unique(interpro_description)), collapse = ", "))
interpro <- bm2wb(df)
```


## Merge and save

```{r save}
df <- cbind(basic,
            entrezGeneId,
            refseq,
            uniprot,
         #! homology,
            geneOntology,
            interpro)
source("geneIdRows.R")
ensembl <- cleanCells(df)
devtools::use_data(ensembl, overwrite = TRUE)
warnings()
```
