---
title: "RNAi annotations"
date: "`r BiocStyle::doc_date()`"
---

```{r setup, include=FALSE}
knitr::opts_knit$set(root.dir = rprojroot::find_package_root_file())
knitr::opts_chunk$set(cache = TRUE)
```

[WORFDB]: http://worfdb.dfci.harvard.edu
[WormBase]: http://www.wormbase.org



# [WormBase][]

```{r wormbaseOligo}
wormbaseOligo <- "pcr_product2gene" %>%
    wormbaseAnnotationFile %>%
    readr::read_tsv(., col_names = c("oligo", "gene")) %>%
    dplyr::mutate(gene = stringr::str_extract(gene, "WBGene\\d{8}"))
devtools::use_data(wormbaseOligo, overwrite = TRUE)
```



# [WORFDB][]

```{r worfdb_functions}
#' @importFrom httr content GET user_agent
#' @importFrom pbmclapply pbmclapply
#' @importFrom stats na.omit
worfdbHtml <- function(sequence) {
    sequence <- sequence %>% stats::na.omit(.) %>% unique
    pbmcapply::pbmclapply(seq_along(sequence), function(a) {
        httr::GET(paste0("http://worfdb.dfci.harvard.edu/searchallwormorfs.pl?by=name&sid=",
                         sequence[a]),
                  user_agent = httr::user_agent(ua)) %>%
            httr::content("text")
    }) %>% stats::setNames(., sequence)
}

#' @importFrom pbmclapply pbmclapply
#' @importFrom stringr str_extract_all str_replace str_replace_all
#' @importFrom tibble as_tibble
worfdbData <- function(worfdbHtml) {
    pbmcapply::pbmclapply(seq_along(worfdbHtml), function(a) {
        html <- worfdbHtml[[a]] %>%
            # Remove `<map>` that has other clone information
            # This messes up well identifier matching otherwise
            gsub("<map.+</map>", "", .)
        clone <- html %>%
            stringr::str_extract_all(., "[0-9]{5}@[A-H][0-9]+") %>%
            unlist %>%
            basejump::toStringUnique(.)
        inFrame <- html %>%
            stringr::str_extract_all(., "In Frame.+<font color=black>([NY])</font>") %>%
            unlist %>%
            stringr::str_replace(., "&nbsp;<font color=black>([NY])</font>", "\\1") %>%
            stringr::str_replace_all(., "Y$", TRUE) %>%
            stringr::str_replace_all(., "N$", FALSE) %>%
            basejump::toStringUnique(.)
        sequence <- html %>%
            stringr::str_match_all(., "<A HREF=http://www.wormbase.org/db/seq/sequence\\?name=([A-Za-z0-9\\.]+)>") %>%
            .[[1]] %>% .[, 2] %>%
            # Strip isoform
            gsub("[a-z]$", "", .) %>%
            basejump::toStringUnique(.)
        sequencingInformation <- html %>%
            stringr::str_extract_all(., "OST in ORFeome version.+\\(WS[0-9]+\\)") %>%
            unlist %>%
            basejump::toStringUnique(.)
        primer <- html %>%
            stringr::str_match_all(., "<font color=red><B>([acgt]+)[\n]?</B></font>") %>%
            .[[1]] %>% .[, 2] %>%
            toupper %>%
            toString
        size <- html %>%
            stringr::str_match_all(., "size: &nbsp;([0-9]+)") %>%
            .[[1]] %>% .[, 2] %>%
            toString
        remap <- html %>%
            stringr::str_match_all(., "<TR><TD><A HREF=searchallwormorfs.pl\\?sid=([A-Z0-9]+\\.[0-9]+[a-z]?)>[A-Z0-9]+\\.[0-9]+[a-z]?</A></TD><TD>([0-9]{5}@[A-H][0-9]+)</TD><TD>([0-9]{5}@[A-H][0-9]+)?</TD><TD>(N|Y)</TD><TD>([0-9]+)</TD></TR>") %>%
            .[[1]] %>% .[, 2] %>%
            basejump::toStringUnique(.)
        list <- list(query = names(worfdbHtml)[a],
                     sequence = sequence,
                     clone = clone,
                     sequencingInformation = sequencingInformation,
                     inFrame = inFrame,
                     primer = primer,
                     remap = remap,
                     size = size)
        lapply(list, function(b) {
            tibble::as_tibble(Filter(Negate(is.null), b))
        })
    }) %>%
        dplyr::bind_rows(.) %>%
        basejump::wash(.) %>%
        dplyr::arrange(sequence) %>%
        dplyr::filter(!is.na(clone)) %>%
        dplyr::mutate(clone = gsub("@", "", clone),
                      clone = gsub("([A-Z]{1})0", "\\1", clone)) %>%
        dplyr::left_join(gene(.$sequence,
                              format = "sequence",
                              select = "gene"),
                         by = "sequence")
}
```

```{r worfdb}
if (!file.exists("data-raw/worfdb/worfdbHtml.rda")) {
    data(wormbaseGene)
    sequence <- wormbaseGene$sequence
    worfdbHtml <- worfdbHtml(sequence)
    save(worfdbHtml, file = "data-raw/worfdb/worfdbHtml.rda")
} else {
    load("data-raw/worfdb/worfdbHtml.rda")
}
worfdb <- worfdbData(worfdbHtml)

if (!file.exists("data-raw/worfdb/worfdbHtmlRemap.rda")) {
    # Example: H15N14.1
    sequenceRemap <- worfdb %>%
        dplyr::filter(!is.na(remap)) %>%
        .$remap %>%
        toString %>%
        stringr::str_split(., ", ") %>%
        unlist
    worfdbHtmlRemap <- worfdbHtml(sequenceRemap)
    save(worfdbHtmlRemap, file = "data-raw/worfdb/worfdbHtmlRemap.rda")
} else {
    load("data-raw/worfdb/worfdbHtmlRemap.rda")
}
worfdbRemap <- worfdbData(worfdbHtmlRemap)

worfdb <- dplyr::bind_rows(worfdb, worfdbRemap) %>%
    dplyr::filter(is.na(remap)) %>%
    dplyr::select(-c(query, remap)) %>%
    dplyr::distinct(.) %>%
    dplyr::arrange(clone)
devtools::use_data(worfdb, overwrite = TRUE)

# Check for duplicates
dupes <- dplyr::filter(worfdb, clone %in% unique(.[["clone"]][duplicated(.[["clone"]])]))
```

```{r cherrypick}
cherrypick <- "data-raw/cherrypick.xlsx" %>%
    readxl::read_excel(.) %>%
    dplyr::mutate(clone = paste0(library,
                                 plateNum,
                                 plateRow,
                                 plateCol)) %>%
    dplyr::filter(!is.na(sequence)) %>%
    dplyr::select(clone, sequence) %>%
    dplyr::arrange(clone) %>%
    dplyr::left_join(gene(.$sequence, format = "sequence"), by = "sequence")
devtools::use_data(cherrypick, overwrite = TRUE)
```

```{r sourcebioscience, eval=FALSE}
fileLocal <- "data-raw/sourcebioscience.xlsx"
download.file("http://www.us.lifesciences.sourcebioscience.com/media/381254/C.%20elegans%20Database%202012.xlsx", fileLocal)
chromosomes <- c("I", "II", "III", "IV", "V", "X")
list <- parallel::mclapply(seq_along(chromosomes), function(a) {
    # Note that the first sheet contains notes, so \code{i + 1}
    readxl::read_excel(fileLocal,
                       sheet = a + 1,
                       col_types = rep("text", 8))
})
input <- dplyr::bind_rows(list) %>%
    basejump::setNamesCamel(.) %>%
    dplyr::filter(!grepl("mismatch", extraInfo)) %>%
    dplyr::rename(genePair = genePairsName,
                  clone384 = sourceBioscienceLocation) %>%
    dplyr::mutate(clone96 = paste0(plate, well),
                  clone96 = gsub("([A-Z]{1})0", "\\1", clone96),
                  clone384 = gsub("-", "", clone384),
                  clone384 = gsub("([A-Z]{1})0", "\\1", clone384),
                  fwdPrimerSeq = toupper(fwdPrimerSeq),
                  revPrimerSeq = toupper(revPrimerSeq)) %>%
    dplyr::select(-c(chrom, plate, well)) %>%
    dplyr::arrange(genePair, clone384) %>%
    dplyr::select(noquote(order(names(.))))
nrow(input)

matched <- list()

# Match `genePair` against `sequence` with `gene()` function:
matched$sequence <- input %>%
    dplyr::mutate(sequence = gsub("^([A-Z0-9]+\\.[0-9]+)[a-z]$", "\\1", genePair)) %>%
    dplyr::left_join(gene(.$sequence,
                          format = "sequence",
                          select = "gene"),
                     by = "sequence") %>%
    dplyr::filter(!is.na(gene))
unmatched <- input %>%
    dplyr::filter(!genePair %in% matched[["sequence"]][["genePair"]])
nrow(matched$sequence)
nrow(unmatched)

# Match by `sjj_` oligo:
data(wormbaseOligo)
matched$oligo <- unmatched %>%
    dplyr::mutate(oligo = paste0("sjj_", genePair)) %>%
    dplyr::left_join(wormbaseOligo, by = "oligo") %>%
    dplyr::filter(!is.na(gene))
unmatched <- unmatched %>%
    dplyr::filter(!genePair %in% matched[["oligo"]][["genePair"]])
nrow(matched$oligo)
nrow(unmatched)

# Match by `sjj2_` oligo:
matched$oligo2 <- unmatched %>%
    dplyr::mutate(oligo = paste0("sjj2_", genePair)) %>%
    dplyr::left_join(wormbaseOligo, by = "oligo") %>%
    dplyr::filter(!is.na(gene))
unmatched <- unmatched %>%
    dplyr::filter(!genePair %in% matched[["oligo2"]][["genePair"]])
nrow(matched$oligo2)
nrow(unmatched)

sourcebioscience <- dplyr::bind_rows(matched, unmatched)
devtools::use_data(sourcebioscience, overwrite = TRUE)
rm(chromosomes,
   fileLocal,
   list,
   input,
   matched,
   unmatched)
```

```{r rnai}
dataRaw(c("cherrypick", "sourcebioscience", "worfdb"))
cherrypick <- cherrypick %>%
    dplyr::select(gene, sequence, clone) %>%
    dplyr::filter(!is.na(gene)) %>%
    dplyr::rename(genePair = sequence)
sourcebioscience <- sourcebioscience %>%
    dplyr::rowwise(.) %>%
    dplyr::mutate(clone = toString(c(clone384, clone96))) %>%
    dplyr::select(gene, genePair, clone)
worfdb <- worfdb %>%
    dplyr::select(gene, sequence, clone) %>%
    dplyr::rename(genePair = sequence)
rnai <- dplyr::bind_rows(cherrypick,
                         sourcebioscience,
                         worfdb) %>%
    dplyr::filter(!is.na(gene)) %>%
    dplyr::group_by(gene) %>%
    basejump::toStringSummarize(.) %>%
    dplyr::left_join(gene(.$gene, select = "biotype"), by = "gene") %>%
    dplyr::select(noquote(order(names(.))))
rm(cherrypick, sourcebioscience, worfdb)
```
