---
title: "RNAi annotations"
date: "`r BiocStyle::doc_date()`"
---

```{r setup, include=FALSE}
knitr::opts_knit$set(root.dir = rprojroot::find_package_root_file())
knitr::opts_chunk$set(cache = TRUE)
```

```{r wormbaseOligo}
wormbaseOligo <- wormbaseAnnotationFile("pcr_product2gene") %>%
    readr::read_tsv(., col_names = c("oligo", "gene")) %>%
    dplyr::mutate(gene = stringr::str_extract(gene, "WBGene\\d{8}"))
devtools::use_data(wormbaseOligo, overwrite = TRUE)
```

```{r worfdb}
# Functions ====
worfdbHtml <- function(sequence) {
    sequence <- sequence %>% na.omit %>% unique
    pbmcapply::pbmclapply(seq_along(sequence), function(a) {
        GET(paste0("http://worfdb.dfci.harvard.edu/searchallwormorfs.pl?by=name&sid=",
                   sequence[a]),
            user_agent = user_agent(ua)) %>%
            content("text")
    }) %>% set_names(sequence)
}

worfdbData <- function(worfdbHtml) {
    pbmcapply::pbmclapply(seq_along(worfdbHtml), function(a) {
        html <- worfdbHtml[[a]] %>%
            # Remove map that has other clones
            # This messes up well identifier matching otherwise
            gsub("<map.+</map>", "", .)
        clone <- html %>%
            str_extract_all("[0-9]{5}@[A-H][0-9]+") %>%
            unlist %>%
            toStringUnique
        inFrame <- html %>%
            str_extract_all("In Frame.+<font color=black>([NY])</font>") %>%
            unlist %>%
            str_replace("&nbsp;<font color=black>([NY])</font>", "\\1") %>%
            str_replace_all("Y$", TRUE) %>%
            str_replace_all("N$", FALSE) %>%
            toStringUnique
        sequence <- html %>%
            str_match_all("<A HREF=http://www.wormbase.org/db/seq/sequence\\?name=([A-Za-z0-9\\.]+)>") %>%
            .[[1]] %>% .[, 2] %>%
            # Strip isoform
            gsub("[a-z]$", "", .) %>%
            toStringUnique
        sequencingInformation <- html %>%
            str_extract_all("OST in ORFeome version.+\\(WS[0-9]+\\)") %>%
            unlist %>%
            toStringUnique
        primer <- html %>%
            str_match_all("<font color=red><B>([acgt]+)[\n]?</B></font>") %>%
            .[[1]] %>% .[, 2] %>%
            toupper %>%
            toString
        size <- html %>%
            str_match_all("size: &nbsp;([0-9]+)") %>%
            .[[1]] %>% .[, 2] %>%
            toString
        remap <- html %>%
            str_match_all("<TR><TD><A HREF=searchallwormorfs.pl\\?sid=([A-Z0-9]+\\.[0-9]+[a-z]?)>[A-Z0-9]+\\.[0-9]+[a-z]?</A></TD><TD>([0-9]{5}@[A-H][0-9]+)</TD><TD>([0-9]{5}@[A-H][0-9]+)?</TD><TD>(N|Y)</TD><TD>([0-9]+)</TD></TR>") %>%
            .[[1]] %>% .[, 2] %>%
            toStringUnique
        list <- list(query = names(worfdbHtml)[a],
                     sequence = sequence,
                     clone = clone,
                     sequencingInformation = sequencingInformation,
                     inFrame = inFrame,
                     primer = primer,
                     remap = remap,
                     size = size)
        lapply(list, function(b) {
            as_tibble(Filter(Negate(is.null), b))
        })
    }) %>%
        bind_rows %>%
        wash %>%
        arrange(sequence) %>%
        filter(!is.na(clone)) %>%
        mutate(clone = gsub("@", "", clone),
               clone = gsub("([A-Z]{1})0", "\\1", clone)) %>%
        left_join(gene(.$sequence, format = "sequence", select = "gene"),
                  by = "sequence")
}

if (!file.exists("data-raw/worfdb/worfdbHtml.rda")) {
    data(wormbaseGene)
    sequence <- wormbaseGene$sequence
    worfdbHtml <- worfdbHtml(sequence)
    save(worfdbHtml, file = "data-raw/worfdb/worfdbHtml.rda")
} else {
    load("data-raw/worfdb/worfdbHtml.rda")
}
worfdb <- worfdbData(worfdbHtml)

if (!file.exists("data-raw/worfdb/worfdbHtmlRemap.rda")) {
    # Example: H15N14.1
    sequenceRemap <- worfdb %>%
        filter(!is.na(remap)) %>%
        .$remap %>% toString %>%
        str_split(", ") %>%
        unlist
    worfdbHtmlRemap <- worfdbHtml(sequenceRemap)
    save(worfdbHtmlRemap, file = "data-raw/worfdb/worfdbHtmlRemap.rda")
} else {
    load("data-raw/worfdb/worfdbHtmlRemap.rda")
}
worfdbRemap <- worfdbData(worfdbHtmlRemap)

worfdb <- bind_rows(worfdb, worfdbRemap) %>%
    filter(is.na(remap)) %>%
    select(-c(query, remap)) %>%
    distinct %>%
    arrange(clone)
use_data(worfdb, overwrite = TRUE)

# Check for duplicates
dupes <- worfdb %>%
    filter(clone %in% unique(.[["clone"]][duplicated(.[["clone"]])]))
```

```{r cherrypick}
cherrypick <- read_excel("data-raw/cherrypick.xlsx") %>%
    mutate(clone = paste0(library,
                          plateNum,
                          plateRow,
                          plateCol)) %>%
    filter(!is.na(sequence)) %>%
    select(clone, sequence) %>%
    arrange(clone) %>%
    left_join(gene(.$sequence, format = "sequence"), by = "sequence")
use_data(cherrypick, overwrite = TRUE)
```

```{r sourcebioscience, eval=FALSE, include=FALSE}
fileLocal <- "data-raw/sourcebioscience.xlsx"
if (!file.exists(fileLocal)) {
    download.file("http://www.us.lifesciences.sourcebioscience.com/media/381254/C.%20elegans%20Database%202012.xlsx",
                  fileLocal)
}
chromosomes <- c("I", "II", "III", "IV", "V", "X")
list <- mclapply(seq_along(chromosomes), function(a) {
    # Note that the first sheet contains notes, so \code{i + 1}
    read_excel(fileLocal, sheet = a + 1, col_types = rep("text", 8))
})
input <- bind_rows(list) %>%
    setNamesCamel %>%
    filter(!grepl("mismatch", extraInfo)) %>%
    rename(genePair = genePairsName,
           clone384 = sourceBioscienceLocation) %>%
    mutate(clone96 = paste0(plate, well),
           clone96 = gsub("([A-Z]{1})0", "\\1", clone96),
           clone384 = gsub("-", "", clone384),
           clone384 = gsub("([A-Z]{1})0", "\\1", clone384),
           fwdPrimerSeq = toupper(fwdPrimerSeq),
           revPrimerSeq = toupper(revPrimerSeq)) %>%
    select(-c(chrom, plate, well)) %>%
    arrange(genePair, clone384) %>%
    select(noquote(order(names(.))))
nrow(input)

matched <- list()

# Match `genePair` against `sequence` with `gene()` function:
matched$sequence <- input %>%
    mutate(sequence = gsub("^([A-Z0-9]+\\.[0-9]+)[a-z]$", "\\1", genePair)) %>%
    left_join(gene(.$sequence,
                   format = "sequence",
                   select = "gene"),
              by = "sequence") %>%
    filter(!is.na(gene))
unmatched <- input %>%
    filter(!genePair %in% matched[["sequence"]][["genePair"]])
nrow(matched$sequence)
nrow(unmatched)

# Match by `sjj_` oligo:
data(wormbaseOligo)
matched$oligo <- unmatched %>%
    mutate(oligo = paste0("sjj_", genePair)) %>%
    left_join(wormbaseOligo, by = "oligo") %>%
    filter(!is.na(gene))
unmatched <- unmatched %>%
    filter(!genePair %in% matched[["oligo"]][["genePair"]])
nrow(matched$oligo)
nrow(unmatched)

# Match by `sjj2_` oligo:
matched$oligo2 <- unmatched %>%
    mutate(oligo = paste0("sjj2_", genePair)) %>%
    left_join(wormbaseOligo, by = "oligo") %>%
    filter(!is.na(gene))
unmatched <- unmatched %>%
    filter(!genePair %in% matched[["oligo2"]][["genePair"]])
nrow(matched$oligo2)
nrow(unmatched)

sourcebioscience <- bind_rows(matched, unmatched)
use_data(sourcebioscience, overwrite = TRUE)
rm(chromosomes,
   fileLocal,
   list,
   input,
   matched,
   unmatched)
```

```{r rnai}
dataRaw(c("cherrypick", "sourcebioscience", "worfdb"))
cherrypick <- cherrypick %>%
    dplyr::select(gene, sequence, clone) %>%
    filter(!is.na(gene)) %>%
    rename(genePair = sequence)
sourcebioscience <- sourcebioscience %>%
    rowwise %>%
    mutate(clone = toString(c(clone384, clone96))) %>%
    select(gene, genePair, clone)
worfdb <- worfdb %>%
    select(gene, sequence, clone) %>%
    rename(genePair = sequence)
rnai <- bind_rows(cherrypick,
                  sourcebioscience,
                  worfdb) %>%
    filter(!is.na(gene)) %>%
    group_by(gene) %>%
    toStringSummarize %>%
    left_join(gene(.$gene, select = "biotype"), by = "gene") %>%
    select(noquote(order(names(.))))
rm(cherrypick, sourcebioscience, worfdb)
```
