---
title: "Ahringer RNAi library annotations"
output: md_document
---

```{r setup, include=FALSE}
knitr::opts_knit$set(root.dir = normalizePath("../"))
```

```{r, include=FALSE}
source("data-raw/setup.R")
```

Use the annotation data from [Source Bioscience](http://www.us.lifesciences.sourcebioscience.com/clone-products/non-mammalian/c-elegans/c-elegans-rnai-library/):

Up-to-date mapping of Julie Ahringer RNAi library clones to current WormBase gene models:
ftp://caltech.wormbase.org/pub/annots/rnai/

```{r ahringer}
ahringer <- list()
if (!file.exists("data-raw/ahringer.xlsx")) {
download.file("http://www.us.lifesciences.sourcebioscience.com/media/381254/C.%20elegans%20Database%202012.xlsx",
              "data-raw/ahringer.xlsx")
}

chromosomes <- c("I", "II", "III", "IV", "V", "X")
for (i in 1:length(chromosomes)) {
  df <- readxl::read_excel("data-raw/ahringer.xlsx", sheet = i + 1)  # First sheet contains notes
  
  colnames(df) <- seqcloudr::camel(colnames(df))
  df$extraInfo <- NULL
  names(df)[names(df) == "genePairsName"] <- "genePair"
  names(df)[names(df) == "sourceBioscienceLocation"] <- "sourceBioscience384"
  
  # Fix primer sequences
  df$fwdPrimerSeq <- tolower(df$fwdPrimerSeq)
  df$revPrimerSeq <- tolower(df$revPrimerSeq)
  
  #! df$plate <- stringr::str_pad(df$plate, 3, pad = "0")
  df$ahringer96 <- do.call(paste, c(df[, c("chrom", "plate", "well")], sep = "-"))
  df$ahringer96[1]
  df$ahringer96 <- gsub("-([A-Z]{1}[0-9]{2})$", "@\\1", df$ahringer96)
  df$ahringer96[1]

  df$historical <- paste0("JA:", df$genePair)
  
  name <- paste0("chr", chromosomes[i])
  assign(name, df)
}
rm(df, i)
list <- list(chrI, chrII, chrIII, chrIV, chrV, chrX)
df <- data.frame(do.call("rbind", list))

ahringerRawData <- df
devtools::use_data(ahringerRawData, overwrite = TRUE)
```

```{r annotations}
# WormBase RESTful API request for RNAi annotations
if (file.exists("data/ahringerRest.rda")) {
  data(ahringerRest)
} else {
  # Split up to prevent curl memory error
  ahringerRest1 <- historical2wbrnai(ahringerRawData$historical[00001:05000])
  ahringerRest2 <- historical2wbrnai(ahringerRawData$historical[05001:10000])
  ahringerRest3 <- historical2wbrnai(ahringerRawData$historical[10001:15000])
  ahringerRest4 <- historical2wbrnai(ahringerRawData$historical[15001:nrow(ahringerRawData)])
  ahringerRest <- rbind(ahringerRest1, ahringerRest2, ahringerRest3, ahringerRest4)
  devtools::use_data(ahringerRest, overwrite = TRUE)
}

# Oligo to Gene ID
if (file.exists("data/oligo2geneId.rda")) {
  data(oligo2geneId)
} else {
  source("data-raw/oligo2geneId.R")
}

data(ahringerRawData)
df <- cbind(ahringerRawData, ahringerRest)
df <- df[, unique(names(df))]
df <- merge(df, oligo2geneId, by = "oligo", all.x = TRUE)

oligoNoMatch <- dplyr::filter(df, is.na(df$geneId))
oligoNoMatch$geneId <- NULL

df <- dplyr::filter(df, !is.na(df$geneId))
df <- merge(df, gene(df$geneId, output = "simple"), by = "geneId", all.x = TRUE)
oligoMatch <- df

# Strip isoforms from genePair search
genePairQuery <- oligoNoMatch$genePair
genePairQuery[1:1000]
genePairQuery <- gsub("([0-9]+)[a-z]$", "\\1", genePairQuery)
genePairQuery[1:1000]
genePairQuery <- gsub("\\.([0-9]+)\\.[0-9]+$", ".\\1", genePairQuery)
df2 <- merge(oligoNoMatch,
             gene(genePairQuery, format = "orf", output = "simple"),
             by.x = "genePair", by.y = "orf",
             all.x = TRUE)

df2 <- merge(df2, gene(df$genePair, format = "orf", output = "simple"), all.x = TRUE)

ahringerData <- df
devtools::use_data(ahringerData, overwrite = TRUE)
```
