---
title: "Gene annotations"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_knit$set(root.dir = normalizePath("../"))
```

```{r, include=FALSE}
source("data-raw/setup.R")
```

# WormBase

## WormBase Gene Identifiers

```{r wormbaseGeneId}
geneIds <- wormbaseFile("geneIDs")
df <- readr::read_csv(geneIds, col_names = FALSE, na = "")
df[[1]] <- NULL
colnames(df) <- c("geneId", "orf", "publicName", "status")
geneIds <- df

# Other identifiers
geneOtherIds <- wormbaseFile("geneOtherIDs")
file <- readr::read_file(geneOtherIds)

# Take out dead or live status, we have this already from geneIds
file <- gsub("\t(Dead|Live)", "", file, perl = TRUE)

# Take the tabs out for gene list
file <- gsub("\t", ", ", file)

# Add tab back in to separate geneId for row names
file <- gsub("WBGene([0-9]+), ", "WBGene\\1\t", file, perl = TRUE)

# Warnings here mean there are no other IDs for that row
# (e.g. expected: 2 columns, actual: 1 columns)
df <- readr::read_tsv(file, col_names = FALSE)

colnames(df) <- c("geneId", "geneOtherIds")
geneOtherIds <- df

df <- cbind(geneIds, geneOtherIds)
df <- df[, !duplicated(colnames(df))]
wormbaseGeneId <- df
devtools::use_data(wormbaseGeneId, overwrite = TRUE)
rm(list = setdiff(ls(), lsf.str()))
```


## WormBase Functional Descriptions

```{r wormbaseDescription}
descriptions <- wormbaseFile("functional_descriptions")
df <-
  readr::read_delim(
    descriptions,
    delim = "\t",
    col_names = FALSE,
    skip = 4,
    # Column headers not properly tabbed
    na = c("", "none available", "not known")
  )

# Fix colnames
header <- readr::read_lines(descriptions, n_max = 4)
header <- header[4]
header <- strsplit(header, " ")
colnames(df) <- header[[1]]
rm(header)

# Convert colnames to camel
colnames(df) <- seqcloudr::camel(colnames(df))
names(df)[names(df) == "molecularName"] <- "orf"

# Remove rows without an identifier (WormBase bug?)
df <- df[grepl("^WBGene[0-9]{8}$", df$geneId), ]

# Duplicates
df$orf <- NULL
df$publicName <- NULL

wormbaseDescription <- df
devtools::use_data(wormbaseDescription, overwrite = TRUE)
rm(list = setdiff(ls(), lsf.str()))
```


## WormBase RNAi Phenotypes

```{r wormbaseRnaiPhenotypes}
if (!file.exists("data-raw/rnai_phenotypes.tsv")) {
  dir <- "ftp://ftp.wormbase.org/pub/wormbase/releases/current-production-release/ONTOLOGY/"
  ls <- RCurl::getURL(dir, dirlistonly = TRUE)
  ls <- strsplit(ls, "\n")
  ls <- ls[[1]]
  grep <- grep("rnai_phenotypes_quick", ls, value = TRUE)
  file <- paste0(dir, grep)
  file
  download.file(file, file.path("data-raw", "rnai_phenotypes.tsv"))
}
df <- readr::read_tsv("data-raw/rnai_phenotypes.tsv", col_names = FALSE)
colnames(df) <- c("geneId", "orf", "unsorted")

list <- parallel::mclapply(seq(along = rownames(df)), function(x) {
  split <- strsplit(as.character(df[x, "unsorted"]), ", ")
  vec <- split[[1]]
  vec <- unique(vec)
  vec <- sort(vec)
  paste(vec, collapse = ", ")
})
resorted <- unlist(list)

# Add the sorted phenotypes back
df <- cbind(df, resorted)

df$orf <- NULL
df$unsorted <- NULL

names(df)[names(df) == "resorted"] <- "rnaiPhenotypes"
wormbaseRnaiPhenotypes <- df
devtools::use_data(wormbaseRnaiPhenotypes, overwrite = TRUE)
rm(list = setdiff(ls(), lsf.str()))
```


## WormBase Orthologs

```{r wormbaseOrthologs}
orthologs <- wormbaseFile("orthologs")
file <- readr::read_file(orthologs)
file <- gsub("\t", " | ", file)
file <- gsub("\n", " // ", file)
file <- gsub("= // ", "\n", file)
file <- gsub(" //  // ", "\t", file)
df <- readr::read_tsv(file, comment = "#", col_names = FALSE)
id <- df[[1]]
head(id)
id <- as.data.frame(stringr::str_split_fixed(id, " \\| ", 2))
head(id)
bind <- cbind(id, df[[2]])
bind[[2]] <- NULL
colnames(bind) <- c("geneId", "orthologs")
orthologsInput <- bind

hsapiensAll <- list()
hsapiensId <- vector()
hsapiensName <- vector()
df <- orthologsInput
# This is inefficient -- work on parallel version
for (i in 1:nrow(df)) {
  gene <- rownames(df)[i]
  split1 <- strsplit(as.character(df[i, "orthologs"]), " // ")
  hsapiensAll[[gene]] <- split1[[1]][grepl("Homo sapiens", split1[[1]])]
  if (length(hsapiensAll[[gene]]) > 0) {
    names <- names(hsapiensAll[[gene]])
    id <- vector()
    name <- vector()
    for (j in 1:length(hsapiensAll[[gene]])) {
      split2 <- strsplit(as.character(hsapiensAll[[gene]][j]), " \\| ")
      id <- append(id, split2[[1]][2])
      name <- append(name, split2[[1]][3])
    }
    id <- unique(id)
    id <- sort(id)
    id <- paste(id, collapse = ", ")
    hsapiensId[i] <- id
    name <- unique(name)
    name <- sort(name)
    name <- paste(name, collapse = ", ")
    hsapiensName[i] <- name
  } else {
    hsapiensId[i] <- NA
    hsapiensName[i] <- NA
  }
}

df <- cbind(orthologsInput, hsapiensName, hsapiensId)
df$orthologs <- NULL

df <- rename(df,
             hsapiensEnsemblGeneName = hsapiensName,
             hsapiensEnsemblGeneId = hsapiensId)

wormbaseOrthologs <- df
devtools::use_data(wormbaseOrthologs, overwrite = TRUE)
rm(list = setdiff(ls(), lsf.str()))
```


## WormBase Best BLASTP Hits

```{r wormbaseBlastp}
# Get the highest match for each peptide
blastpScores <- wormbaseFile("best_blast_hits")
input <- readr::read_csv(blastpScores, col_names = FALSE)
df <- input[, c(1, 4, 5)]
colnames(df) <- c("wormpepId", "ensemblPeptideId", "eValue")

# Filter for only ENSEMBL info
df <- df[grepl("ENSEMBL", df$ensemblPeptideId), ]

# Remove "ENSEMBL:" from column
df$ensemblPeptideId <- substr(df$ensemblPeptideId, 9, 23)

# Sort by E value to get the highest confidence BLASTP match
df <- arrange(df, wormpepId, eValue)

# Now remove duplicates, will which eliminate the lower confidence entries
df <- distinct(df)

blastpScores <- df

# Wormpep IDs are used for BLASTP matching:
if (!file.exists("data-raw/wormpep.txt")) {
  dir <- "ftp://ftp.wormbase.org/pub/wormbase/releases/current-production-release/species/c_elegans/PRJNA13758/"
  ls <- RCurl::getURL(dir, dirlistonly = TRUE)
  ls <- strsplit(ls, "\n")
  ls <- ls[[1]]
  grep <- grep("wormpep_package", ls, value = TRUE)
  file <- paste0(dir, grep)
  download.file(file, "wormpep.tar.gz")
  
  # Extract wormpep.table specifically -- works on macOS
  system("tar -xzf wormpep.tar.gz wormpep.table*")
  file <- list.files(pattern = "wormpep.table")
  file.rename(file, file.path("data-raw", "wormpep.txt"))
  file.remove("wormpep.tar.gz")
}

input <- readr::read_lines("data-raw/wormpep.txt")
input <- strsplit(input, "\n")
wormpep <- parallel::mclapply(input, function(x) {
  x <- gsub("^.*\t(CE[0-9]+\tWBGene[0-9]+).*$", "\\1", x, perl = T)
  y <- strsplit(x, "\t")
  x <- y[[1]]
})
head(wormpep)
df <- data.frame(do.call("rbind", wormpep))
names(df) <- c("wormpepId", "geneId")
wormpepId <- df

# Bind the WormBase Gene ID
df <- merge(blastpScores, wormpepId, by = "wormpepId", all = TRUE)

# Subset only the top blastp with ensembl match
df <- df[order(df$geneId, df$eValue, df$wormpepId), ]
df <- df[!duplicated(df$geneId), ]
df <- df[!is.na(df$ensemblPeptideId), ]
df <- df[, unique(colnames(df))]
blastpGeneId <- df

input <- as.vector(blastpGeneId$ensemblPeptideId)
mart <- biomaRt::useMart("ensembl", "hsapiens_gene_ensembl")
biomartOptions <- biomaRt::listAttributes(mart)
martOutput <-
  biomaRt::getBM(mart = mart,
                 filters = "ensembl_peptide_id",
                 values = input,
                 attributes = c("ensembl_peptide_id",
                                "ensembl_gene_id",
                                "external_gene_name",
                                "description")
  )
df <- martOutput
colnames(df)[colnames(df) == "external_gene_name"] <- "ensembl_gene_name"
colnames(df)[colnames(df) == "description"] <- "ensembl_description"
colnames(df) <- seqcloudr::camel(colnames(df))
biomartHsapiens <- df

# Final merge
wormbaseBlastp <- merge(blastpGeneId,
                        biomartHsapiens,
                        by = "ensemblPeptideId",
                        all = TRUE)
devtools::use_data(wormbaseBlastp, overwrite = TRUE)
rm(list = setdiff(ls(), lsf.str()))
```



# Ensembl

## Connect to Biomart

* `entrezgene` = Entrez identifier
* `external_gene_name` = Ensembl public name
* `wormbase_locus` = WormBase public name

Use `wormbase_gene_seq_name` for clean sequence identifier.

```{r ensembl connect}
mart <- biomaRt::useMart("ensembl", "celegans_gene_ensembl")
biomartOptions <- biomaRt::listAttributes(mart)
```


## Simple gene length info

```{r ensembl basic}
basic <-
  biomaRt::getBM(mart = mart,
                 attributes = c("ensembl_gene_id",
                                "gene_biotype",
                                "chromosome_name",
                                "start_position",
                                "end_position",
                                "strand",
                                "description")
  )
```


## Identifiers

```{r ensembl otherIds}
bm1 <- biomaRt::getBM(mart = mart,
                     attributes = c("ensembl_gene_id",
                                    "entrezgene",
                                    "kegg_enzyme"))
bm2 <- biomaRt::getBM(mart = mart,
                     attributes = c("ensembl_gene_id",
                                    "refseq_mrna",
                                    "refseq_ncrna"))
bm3 <- biomaRt::getBM(mart = mart,
                     attributes = c("ensembl_gene_id",
                                    "uniprot_sptrembl",
                                    "uniprot_swissprot"))
df <-
  Reduce(function(...) {
      merge(..., by = "ensembl_gene_id", all = TRUE)
    },
    list(bm1, bm2, bm3)
  )
otherIds <-
  df %>% group_by(ensembl_gene_id) %>%
  summarize(
    entrez_gene_id = paste(sort(unique(entrezgene)), collapse = ", "),
    kegg_enzyme = paste(sort(unique(kegg_enzyme)), collapse = ", "),
    refseq_mrna = paste(sort(unique(refseq_mrna)), collapse = ", "),
    refseq_ncrna = paste(sort(unique(refseq_ncrna)), collapse = ", "),
    uniprot_sptrembl = paste(sort(unique(uniprot_sptrembl)), collapse = ", "),
    uniprot_swissprot = paste(sort(unique(uniprot_swissprot)), collapse = ", ")
  )
```

## Homology

```{r ensembl homology, eval=FALSE, include=FALSE}
bm <- biomaRt::getBM(mart = mart,
                     attributes = c("ensembl_gene_id",
                                    "hsapiens_homolog_ensembl_gene"))
homology <- bm %>% group_by(ensembl_gene_id) %>%
  summarize(
    hsapiens_homolog_ensembl_gene = paste(sort(unique(hsapiens_homolog_ensembl_gene)), collapse = ", ")
  )
```


## Gene Ontology

```{r ensembl geneOntology}
bm <- biomaRt::getBM(mart = mart,
                     attributes = c("ensembl_gene_id",
                                    "go_id",
                                    "name_1006"))
geneOntology <- bm %>% group_by(ensembl_gene_id) %>%
  summarize(
    gene_ontology_name = paste(sort(unique(name_1006)), collapse = ", "),
    gene_ontology_id = paste(sort(unique(go_id)), collapse = ", ")
  )
```


## Interpro

```{r ensembl interpro}
bm <- biomaRt::getBM(mart = mart,
                     attributes = c("ensembl_gene_id",
                                    "interpro",
                                    "interpro_description"))
interpro <- bm %>% group_by(ensembl_gene_id) %>%
  summarize(
    interpro_id = paste(sort(unique(interpro)),collapse = ", "),
    interpro_description = paste(sort(unique(interpro_description)), collapse = ", ")
  )
```


## Merge and Save

```{r ensembl save}
df <- 
  Reduce(function(...) {
      merge(..., by = "ensembl_gene_id", all = TRUE)
    },
    list(basic,
         otherIds,
         geneOntology,
         interpro)
  )
df <- df[, !duplicated(colnames(df))]
ensembl <- seqcloudr::cruft(df)
devtools::use_data(ensembl, overwrite = TRUE)
```



# PANTHER

```{r panther}
if (!file.exists("data-raw/panther.txt")) {
  dir <- "ftp://ftp.pantherdb.org/sequence_classifications/current_release/PANTHER_Sequence_Classification_files/"
  ls <- RCurl::getURL(dir, dirlistonly = TRUE)
  ls <- strsplit(ls, "\n")
  ls <- ls[[1]]
  grep <- grep("nematode_worm", ls, value = TRUE)
  file <- paste0(dir, grep)
  download.file(file, file.path("data-raw", "panther.txt"))
}

# Load and set column names
df <- readr::read_delim(file.path("data-raw", "panther.txt"),
                        delim = "\t",
                        col_names = FALSE)
colnames(df) <- c(
  "id",
  "protein",
  "subfamilyId",
  "familyName",
  "subfamilyName",
  "geneOntologyMolecularFunction",
  "geneOntologyBiologicalProcess",
  "geneOntologyCellularComponent",
  "class",
  "pathway"
  )
df$id <- gsub("CAEEL\\|", "", df$id)
df$protein <- NULL
df$subfamilyId <- NULL
pantherInput <- df

# Fix incorrect ID mappings
id <- pantherInput$id
pantherFix <- filter(df, !grepl("=WBGene", id))

## ORF
id <- gsub("Gene=B0303.5\\|",
           "WormBase=WBGene00015127\\|", id)
id <- gsub("Gene=C13B9.1\\|",
           "WormBase=WBGene00015732\\|", id)
id <- gsub("Gene=F23F12.11/F23F12.5\\|",
           "WormBase=WBGene00005075\\|", id)
id <- gsub("Gene=F55A11.4\\|",
           "WormBase=WBGene00010077\\|", id)
id <- gsub("Gene=K02A2.6\\|",
           "WormBase=WBGene00019291\\|", id)
id <- gsub("Gene=MTCE.11\\|",
           "WormBase=WBGene00010959\\|", id)
id <- gsub("Gene=T07D3.8\\|",
           "WormBase=WBGene00020310\\|", id)
id <- gsub("Gene=Y104H12A.1/Y77E11A.5\\|",
           "WormBase=WBGene00022423\\|", id)
id <- gsub("Gene=Y62E10A.11\\|",
           "WormBase=WBGene00014938\\|", id)

## Entrez
id <- gsub("GeneID=176090\\|",
           "WormBase=WBGene00022794\\|", id) # ZK686.4
id <- gsub("GeneID=2565696\\|",
           "WormBase=WBGene00010966\\|", id) # MTCE.34
id <- gsub("GeneID=2565700\\|",
           "WormBase=WBGene00010964\\|", id) # MTCE.26
id <- gsub("GeneID=2565701\\|",
           "WormBase=WBGene00010962\\|", id) # MTCE.23
id <- gsub("GeneID=2565702\\|",
           "WormBase=WBGene00000829\\|", id) # MTCE.21
id <- gsub("GeneID=2565703\\|",
           "WormBase=WBGene00010967\\|", id) # MTCE.35
id <- gsub("GeneID=2565704\\|",
           "WormBase=WBGene00010960\\|", id) # MTCE.12
id <- gsub("GeneID=2565705\\|",
           "WormBase=WBGene00010963\\|", id) # MTCE.25

df <- data.frame(stringr::str_split_fixed(id, "\\|", 2))
colnames(df) <- c("geneId", "uniprotKb")
df$geneId <- gsub("^(EnsemblGenome|WormBase)=", "", df$geneId)
df$uniprotKb <- gsub("^UniProtKB=", "", df$uniprotKb)
head(df)

# Recombine then remove duplicates from df, take out UniProtKB, then add back
df <- cbind(df, pantherInput)

# Now okay to remove
df$id <- NULL

# Convert semicolon to double slash
df <- data.frame(apply(df, c(1, 2), function(x) gsub(";", " // ", x)))

panther <- df
devtools::use_data(panther, overwrite = TRUE)
rm(list = setdiff(ls(), lsf.str()))
```



# Merge gene annotation information

```{r geneInfo}
data(wormbaseGeneId,
     wormbaseDescription,
     wormbaseRnaiPhenotypes,
     wormbaseOrthologs,
     wormbaseBlastp,
     ensembl,
     panther)


# Column fixes
ensembl <- rename(ensembl,
                  geneId = ensembl_gene_id,
                  ensembl_description = description)

names(panther)[3:length(panther)] <- paste("panther", names(panther)[3:length(panther)], sep = "_")

wormbaseBlastp <- wormbaseBlastp %>% select(geneId, wormpepId, everything())
names(wormbaseBlastp) <- gsub("^ensembl", "hsapiensBlastpEnsembl", names(wormbaseBlastp))
wormbaseBlastp$eValue <- NULL

df <- Reduce(function(...) {
      merge(..., by = "geneId", all = TRUE)
    },
    list(wormbaseGeneId,
         wormbaseDescription,
         wormbaseRnaiPhenotypes,
         wormbaseOrthologs,
         wormbaseBlastp,
         ensembl,
         panther)
  )
names(df) <- seqcloudr::camel(names(df))
rownames(df) <- df[[1]]
df <- seqcloudr::cruft(df)
geneInfo <- df
devtools::use_data(geneInfo, overwrite = TRUE)
```
