---
title: "Gene annotations"
author: "Michael J. Steinbaugh"
date: "`r Sys.Date()`"
output:
    rmarkdown::html_vignette:
        css: http://static.steinbaugh.com/css/vignette.css
        self_contained: false
vignette: >
    %\VignetteIndexEntry{gene}
    %\VignetteEngine{knitr::rmarkdown}
    %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(results = "asis", render = pander::pander)
pander::panderOptions("table.style", "rmarkdown")
pander::panderOptions("table.alignment.default", "left")
pander::panderOptions("table.alignment.rownames", "right")
library(worminfo)
```

```{r, include = FALSE}
markdownList <- function(character) {
    string <- sapply(seq_along(character), function(a) {
        paste0("- ", character[a])
    })
    return(cat(paste(string, collapse = "\n")))
}
```

[dplyr]: https://cran.r-project.org/package=dplyr
[ensembl]: http://www.ensembl.org/Caenorhabditis_elegans
[panther]: http://www.pantherdb.org
[tibble]: https://cran.r-project.org/package=tibble
[worminfo]: https://github.com/steinbaugh/worminfo
[wormbase]: http://www.wormbase.org

[*daf-2*]: http://www.wormbase.org/species/c_elegans/gene/WBGene00000898
[*daf-16*]: http://www.wormbase.org/species/c_elegans/gene/WBGene00000912
[*sbp-1*]: http://www.wormbase.org/species/c_elegans/gene/WBGene00004735
[*skn-1*]: http://www.wormbase.org/species/c_elegans/gene/WBGene00004804

The goal of the [worminfo][] package is to quickly and simply provide correct,
up-to-date annotation data for the entire *C. elegans* genome. The underlying
annotations provided by the `gene()` function are automatically sourced from
[WormBase][], [Ensembl][] and [PANTHER][]. These source files are saved at the
[data branch](https://github.com/steinbaugh/worminfo/tree/data) of the [worminfo
GitHub repository][worminfo]. [WormBase][] is updated on a bi-monthly [release
schedule](http://www.wormbase.org/about/release_schedule). If the source data
appears outdated, file an [update
request](https://github.com/steinbaugh/worminfo/issues) on GitHub, and we'll
promptly issue a rebuild.

`gene()` accepts 4 types of identifiers: `gene`, `name`, `sequence` or `class`.
Output from the function is controlled using the `select` parameter. The
function returns a [tibble][] that can be manipulated using the [dplyr][]
package or the `base::merge()` function. We recommend using
`dplyr::left_join()` to bind these annotations to other `data.frames`.


## `identifier` examples

[WormBase][] **gene**:

```{r}
identifier <- c("WBGene00000898", "WBGene00000912", "WBGene00004804")
gene(identifier)
```


**Sequence**, commonly the open reading frame (ORF):

```{r}
c("Y55D5A.5", "R13H8.1", "T19E7.2") %>% gene(format = "sequence")
```


Public **name**:

```{r}
c("daf-2", "daf-16", "skn-1") %>% gene(format = "name")
```


See how these queries all return the same [tibble][] `data.frame`, but with
different sorting.

You can also return all genes in a named **class**. For example, let's get the
current list of *daf* (**D**auer **A**rrest **F**orming) genes:

```{r}
gene("daf", format = "class")
```


You can perform a **keyword** search. For example, let's query all genes that
have been implicated in the regulation of worm lifespan:

```{r}
gene("lifespan",
     format = "keyword",
     select = c("gene", "sequence", "name")) %>% head
```


## `select` options

The `gene()` function is bundled with source data that can be queried using the
`select` parameter. Here's the current list of data annotations that can be
called with `select`:

```{r, echo = FALSE, results = "asis"}
get("geneAnnotation", envir = asNamespace("worminfo")) %>%
    names %>%
    markdownList
```


When `select` is not declared, a simple result is returned:

```{r, echo = FALSE, results = "asis"}
gene(identifier) %>% names %>% markdownList
```


## `gene()` function examples

Basic identifiers for [*skn-1*][] (bZIP transcription factor):

```{r}
gene("skn-1", format = "name") %>% t
```

Select [WormBase][] and [Ensembl][] descriptions for [*sbp-1*][] (bHLH
transcription factor):

```{r}
gene("sbp-1",
     format = "name",
     select = c("descriptionConcise",
                "descriptionProvisional",
                "ensemblDescription")) %>% t
```
