---
title: "WormBase RESTful API"
author: "Michael J. Steinbaugh"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{rnai}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
library(magrittr)
devtools::load_all()
```

[httr]: https://cran.r-project.org/web/packages/httr
[json]: http://json.org
[wormbase]: http://www.wormbase.org
[r]: https://www.r-project.org
[restful]: http://www.wormbase.org/about/userguide/for_developers/API-REST

[*daf-16*]: http://www.wormbase.org/species/c_elegans/gene/WBGene00000912
[*sbp-1*]: http://www.wormbase.org/species/c_elegans/gene/WBGene00004735

[WormBase][] has a [RESTful][] interface enabled on their website that allows researchers to programmatically access data. For queries into [R][], we make this calls as [JSON][] (JavaScript Object Notation) requests using the [httr][] package.

We have defined a bare bones function called `rest()`, which defines a [REST][] call to the [WormBase][] website. Any supported URL in the [WormBase][] [developer user guide](http://www.wormbase.org/about/userguide/for_developers) can be queried with this function.

Here's an example that pulls all the data in the Overview widget for the first gene defined at [WormBase][], `WBGene00000001`:

```{r}
rest("widget/gene/WBGene00000001/overview")
# rest("widget/gene/WBGene00000001/reagent")
rest("widget/gene/WBGene00000001/location")
rest("widget/gene/WBGene00000001/genetics")
rest("widget/gene/WBGene00000001/references")
rest("widget/gene/WBGene00000001/homology")
rest("widget/gene/WBGene00000001/human_diseases")
rest("widget/gene/WBGene00000001/sequences")

# Could use this instead of RNAi phenotypes from WormBaseFTP...
rest("widget/gene/WBGene00000001/phenotypes")
```

Any selectable widget on the WormBase website call be called with this simple function.



# Gene calls

We've also made a number of utility functions that make repeated data queries easier:

## restGeneExternal

```{r}
restGeneExternal("WBGene00000001") %>% t
```

## restGeneOntology

```{r}
restGeneOntology("WBGene00000001") %>% t
```




# RNAi calls

```{r}
rest("widget/rnai/WBRNAi00000001/overview")
```

```{r}
rest("widget/rnai/WBRNAi00000001/reagent")
```

## restRnaiReagent

```{r}
restRnaiReagent("
```

## restRnaiSequence

## restRnaiTargets

## Finding clones by gene identifier

Let's find the locations of [*sbp-1*][] in the [Ahringer][] and [ORFeome][]
libraries. This can be done easily with the `rnai()` function, just search by
**name**:

```{r}
rnai("sbp-1", format = "name") %>% t
```

Alternatively, you can search by [WormBase][] **gene** identifier or
**sequence**, aka open reading frame (ORF). These queries will return the same
result:

```{r, eval = FALSE}
rnai("WBGene00004735", format = "gene")
rnai("Y47D3B.7", format = "sequence")
```


## Mapping RNAi clones to gene identifiers

The `rnai()` function was written to simplify the task of identifying wells of
interest in an RNAi screen. You can simply input the well identifiers by plate
location, and `rnai()` will return up-to-date gene identifiers from [WormBase][].

### [Ahringer][] library

Here's how you map wells from the [Ahringer][] library in 384 well format. Note
that plate location identifiers must include the chromosome in roman numeral
format at the beginning:

```{r}
rnai("III-6C01", library = "ahringer384") %>% t
```

Wells from the [Ahringer][] library in 96 well format are also supported. Since
this library has unique plate numbers, the chromosome prefix is not necessary:

```{r}
rnai("86B01", library = "ahringer96") %>% t
```

### [ORFeome][] library

`rnai()` has mapping support for the [ORFeome][] library, in 96 well format. The
function is flexible, and also supports entries in the historical format (e.g.
`GHR-11010@G06`), or without a separator:

```{r}
rnai("11010G06", library = "orfeome96") %>% t
```
