---
title: "Clone mapping to ORFeome and Ahringer feeding RNAi libraries"
author: "Michael J. Steinbaugh"
date: "`r Sys.Date()`"
output:
    rmarkdown::html_vignette:
        css: http://static.steinbaugh.com/css/vignette.css
        self_contained: false
vignette: >
    %\VignetteIndexEntry{rnai}
    %\VignetteEngine{knitr::rmarkdown}
    %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(results = "asis", render = pander::pander)
pander::panderOptions("table.style", "rmarkdown")
pander::panderOptions("table.alignment.default", "left")
pander::panderOptions("table.alignment.rownames", "right")
library(worminfo)
```

[ahringer]: http://www.us.lifesciences.sourcebioscience.com/clone-products/non-mammalian/c-elegans/c-elegans-rnai-library/
[orfeome]: http://worfdb.dfci.harvard.edu
[orfeome dharmacon]: http://dharmacon.gelifesciences.com/non-mammalian-cdna-and-orf/c.-elegans-rnai/
[orfeome sbs]: http://www.us.lifesciences.sourcebioscience.com/clone-products/non-mammalian/c-elegans/c-elegans-orfeome-version-11/
[wormbase]: http://www.wormbase.org

[*daf-16*]: http://www.wormbase.org/species/c_elegans/gene/WBGene00000912
[*sbp-1*]: http://www.wormbase.org/species/c_elegans/gene/WBGene00004735

RNA interference libraries are commonly used in *C. elegans* research. RNAi
clones are typically referenced by a gene pair or open reading frame (ORF)
sequence, which are often difficult to map to an up-to-date gene identifier. We
have written the `rnai()` function to solve this problem. The function enables a
researcher to search for RNAi clones by a gene identifier, or map an RNAi clone
by plate location and return the up-to-date [WormBase][] gene identifier. The
function also obtains primary and secondary target information from
[WormBase][], which can be informative since many RNAi clones have off-target
effects.

The function has built-in support for the commonly used [Ahringer][ahringer] and
[ORFeome][] RNAi libraries.


## Finding clones by gene identifier

Let's find the locations of [*sbp-1*][] in the [Ahringer][] and [ORFeome][]
libraries. This can be done easily with the `rnai()` function, just search by
**name**:

```{r}
rnai("sbp-1", format = "name") %>% t
```

Alternatively, you can search by [WormBase][] **gene** identifier or
**sequence**, aka open reading frame (ORF). These queries will return the same
result:

```{r, eval = FALSE}
rnai("WBGene00004735", format = "gene")
rnai("Y47D3B.7", format = "sequence")
```


## Mapping RNAi clones to gene identifiers

The `rnai()` function was written to simplify the task of identifying wells of
interest in an RNAi screen. You can simply input the well identifiers by plate
location, and `rnai()` will return up-to-date gene identifiers from [WormBase][].

### [Ahringer][] library

Here's how you map wells from the [Ahringer][] library in 384 well format. Note
that plate location identifiers must include the chromosome in roman numeral
format at the beginning:

```{r}
rnai("III-6C01", library = "ahringer384") %>% t
```

Wells from the [Ahringer][] library in 96 well format are also supported. Since
this library has unique plate numbers, the chromosome prefix is not necessary:

```{r}
rnai("86B01", library = "ahringer96") %>% t
```

### [ORFeome][] library

`rnai()` has mapping support for the [ORFeome][] library, in 96 well format.
Plate position should be entered in the same format as seen on the WORFDB.
However, the function is flexible, and also supports entries in the historical
format (e.g. `GHR-11010@G06`), or without a separator:

```{r}
rnai("GHR-11010@G06", library = "orfeome96") %>% t
rnai("11010G6", library = "orfeome96") %>% t
```
