<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>worminfo • worminfo</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/spacelab/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><!-- mathjax --><script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">worminfo</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/index.html">Articles</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>worminfo</h1>
                        <h4 class="author">Michael J. Steinbaugh</h4>
            
            <h4 class="date">2017-01-19</h4>
          </div>

    
    
<div class="contents">
<p>The goal of the <a href="https://github.com/steinbaugh/worminfo">worminfo</a> package is to quickly and simply provide correct, up-to-date annotation data for the entire <em>C. elegans</em> genome. The underlying annotations provided by the <code><a href="../reference/gene.html">gene()</a></code> function are automatically sourced from <a href="http://www.wormbase.org">WormBase</a>, <a href="http://www.ensembl.org/Caenorhabditis_elegans">Ensembl</a> and <a href="http://www.pantherdb.org">PANTHER</a>. These source files are saved at the <a href="https://github.com/steinbaugh/worminfo/tree/data">data branch</a> of the <a href="https://github.com/steinbaugh/worminfo">worminfo GitHub repository</a>. <a href="http://www.wormbase.org">WormBase</a> is updated on a bi-monthly <a href="http://www.wormbase.org/about/release_schedule">release schedule</a>. If the source data appears outdated, file an <a href="https://github.com/steinbaugh/worminfo/issues">update request</a> on GitHub, and we’ll promptly issue a rebuild.</p>
<p>Current annotation build information:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">get</span>(<span class="st">"build"</span>, <span class="dt">envir =</span> <span class="kw">asNamespace</span>(<span class="st">"worminfo"</span>))</code></pre></div>
<pre><code>## $date
## [1] "2017-01-19"
## 
## $ensembl
## [1] "Ensembl Genes 87"
## 
## $panther
## [1] "PANTHER 11.0"
## 
## $rstats
## [1] "R version 3.3.2 (2016-10-31) running on x86_64-apple-darwin13.4.0"
## 
## $wormbase
## [1] "WormBase WS256"</code></pre>
<div id="gene-identifier-queries" class="section level1">
<h1 class="hasAnchor">
<html><body><a href="#gene-identifier-queries" class="anchor"> </a></body></html>Gene identifier queries</h1>
<p><code><a href="../reference/gene.html">gene()</a></code> accepts 4 types of identifiers: <code>gene</code>, <code>name</code>, <code>sequence</code> or <code>class</code>. Output from the function is controlled using the <code>select</code> parameter. The function returns a <a href="https://cran.r-project.org/package=tibble">tibble</a> that can be manipulated using the <a href="https://cran.r-project.org/package=dplyr">dplyr</a> package or the <code>base::merge()</code> function. We recommend using <code>dplyr::left_join()</code> to bind these annotations to other <code>data.frames</code>.</p>
<ul>
<li>
<code>gene</code>: <a href="http://www.wormbase.org">WormBase</a> <strong>gene</strong>:</li>
<li>
<code>sequence</code>: <a href="http://www.wormbase.org">WormBase</a> <strong>sequence</strong>, aka ORF:</li>
<li>
<code>name</code>: <a href="http://www.wormbase.org">WormBase</a> <strong>name</strong>:</li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/gene.html">gene</a></span>(<span class="st">"WBGene00004804"</span>, <span class="dt">format =</span> <span class="st">"gene"</span>)
<span class="kw"><a href="../reference/gene.html">gene</a></span>(<span class="st">"T19E7.2"</span>, <span class="dt">format =</span> <span class="st">"sequence"</span>)
<span class="kw"><a href="../reference/gene.html">gene</a></span>(<span class="st">"skn-1"</span>, <span class="dt">format =</span> <span class="st">"name"</span>)</code></pre></div>
<p>All 3 of these queries return the same result:</p>
<p>See how these queries all return the same <a href="https://cran.r-project.org/package=tibble">tibble</a> <code>data.frame</code>, but with different sorting.</p>
<p>You can also return all genes in a named <strong>class</strong>. For example, let’s get the current list of <em>daf</em> (<strong>D</strong>auer <strong>A</strong>rrest <strong>F</strong>orming) genes:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/gene.html">gene</a></span>(<span class="st">"daf"</span>, <span class="dt">format =</span> <span class="st">"class"</span>)</code></pre></div>
<pre><code>## # A tibble: 34 × 3
##      name           gene sequence
##     &lt;chr&gt;          &lt;chr&gt;    &lt;chr&gt;
## 1   daf-1 WBGene00000897  F29C4.1
## 2   daf-2 WBGene00000898 Y55D5A.5
## 3   daf-3 WBGene00000899  F25E2.5
## 4   daf-4 WBGene00000900  C05D2.1
## 5   daf-5 WBGene00000901  W01G7.1
## 6   daf-6 WBGene00000902  F31F6.5
## 7   daf-7 WBGene00000903  B0412.2
## 8   daf-8 WBGene00000904 R05D11.1
## 9   daf-9 WBGene00000905  T13C5.1
## 10 daf-10 WBGene00000906  F23B2.4
## # ... with 24 more rows</code></pre>
<p>You can perform a <strong>keyword</strong> search. For example, let’s query all genes that have been implicated in the regulation of worm lifespan:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/gene.html">gene</a></span>(<span class="st">"lifespan"</span>, <span class="dt">format =</span> <span class="st">"keyword"</span>) %&gt;%<span class="st"> </span>head</code></pre></div>
<pre><code>## # A tibble: 6 × 4
##             gene  keyword   sequence   name
##            &lt;chr&gt;    &lt;chr&gt;      &lt;chr&gt;  &lt;chr&gt;
## 1 WBGene00000001 lifespan Y110A7A.10  aap-1
## 2 WBGene00000037 lifespan   Y48B6A.8  ace-3
## 3 WBGene00000038 lifespan   Y48B6A.7  ace-4
## 4 WBGene00000040 lifespan    ZK455.1  aco-1
## 5 WBGene00000041 lifespan   F54H12.1  aco-2
## 6 WBGene00000055 lifespan    F25G6.3 acr-16</code></pre>
<div id="select-options" class="section level2">
<h2 class="hasAnchor">
<html><body><a href="#select-options" class="anchor"> </a></body></html>
<code>select</code> options</h2>
<p>The <code><a href="../reference/gene.html">gene()</a></code> function is bundled with source data that can be queried using the <code>select</code> parameter. Here’s the current list of data annotations that can be called with <code>select</code>:</p>
<ul>
<li>biotype</li>
<li>blastpHsapiensDescription</li>
<li>blastpHsapiensGene</li>
<li>blastpHsapiensName</li>
<li>blastpPeptide</li>
<li>blastpWormpep</li>
<li>chromosome</li>
<li>class</li>
<li>descriptionAutomated</li>
<li>descriptionConcise</li>
<li>descriptionDetailed</li>
<li>descriptionProvisional</li>
<li>endPosition</li>
<li>ensemblDescription</li>
<li>ensemblGeneOntology</li>
<li>gene</li>
<li>interpro</li>
<li>name</li>
<li>orthologHsapiens</li>
<li>otherIdentifier</li>
<li>pantherClass</li>
<li>pantherFamilyName</li>
<li>pantherGeneOntologyBiologicalProcess</li>
<li>pantherGeneOntologyCellularComponent</li>
<li>pantherGeneOntologyMolecularFunction</li>
<li>pantherPathway</li>
<li>pantherSubfamily</li>
<li>pantherSubfamilyName</li>
<li>rnaiPhenotype</li>
<li>sequence</li>
<li>startPosition</li>
<li>status</li>
<li>strand</li>
</ul>
<p>When <code>select</code> is not declared, a simple result is returned:</p>
<ul>
<li>gene</li>
<li>sequence</li>
<li>name</li>
</ul>
</div>
<div id="gene-query-examples" class="section level2">
<h2 class="hasAnchor">
<html><body><a href="#gene-query-examples" class="anchor"> </a></body></html>
<code><a href="../reference/gene.html">gene()</a></code> query examples</h2>
<p>Basic identifiers for <a href="http://www.wormbase.org/species/c_elegans/gene/WBGene00004804"><em>skn-1</em></a> (bZIP transcription factor):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/gene.html">gene</a></span>(<span class="st">"skn-1"</span>, <span class="dt">format =</span> <span class="st">"name"</span>) %&gt;%<span class="st"> </span>t</code></pre></div>
<pre><code>##          [,1]            
## name     "skn-1"         
## gene     "WBGene00004804"
## sequence "T19E7.2"</code></pre>
<p>Select <a href="http://www.wormbase.org">WormBase</a> and <a href="http://www.ensembl.org/Caenorhabditis_elegans">Ensembl</a> descriptions for <a href="http://www.wormbase.org/species/c_elegans/gene/WBGene00004735"><em>sbp-1</em></a> (bHLH transcription factor):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/gene.html">gene</a></span>(<span class="st">"sbp-1"</span>,
     <span class="dt">format =</span> <span class="st">"name"</span>,
     <span class="dt">select =</span> <span class="st">"ensemblDescription"</span>) %&gt;%<span class="st"> </span>t</code></pre></div>
<pre><code>##          [,1]            
## name     "sbp-1"         
## gene     "WBGene00004735"
## sequence "Y47D3B.7"</code></pre>
</div>
</div>
<div id="rnai-clone-identifier-queries" class="section level1">
<h1 class="hasAnchor">
<html><body><a href="#rnai-clone-identifier-queries" class="anchor"> </a></body></html>RNAi clone identifier queries</h1>
<p>RNA interference libraries are commonly used in <em>C. elegans</em> research. RNAi clones are typically referenced by a gene pair or open reading frame (ORF) sequence, which are often difficult to map to an up-to-date gene identifier. We have written the <code><a href="../reference/rnai.html">rnai()</a></code> function to solve this problem. The function enables a researcher to search for RNAi clones by a gene identifier, or map an RNAi clone by plate location and return the up-to-date <a href="http://www.wormbase.org">WormBase</a> gene identifier. The function also obtains primary and secondary target information from <a href="http://www.wormbase.org">WormBase</a>, which can be informative since many RNAi clones have off-target effects.</p>
<p>The function has built-in support for the commonly used <a href="http://www.us.lifesciences.sourcebioscience.com/clone-products/non-mammalian/c-elegans/c-elegans-rnai-library/">Ahringer</a> and <a href="http://worfdb.dfci.harvard.edu">ORFeome</a> RNAi libraries.</p>
<div id="finding-clones-by-gene-identifier" class="section level2">
<h2 class="hasAnchor">
<html><body><a href="#finding-clones-by-gene-identifier" class="anchor"> </a></body></html>Finding clones by gene identifier</h2>
<p>Let’s find the locations of <a href="http://www.wormbase.org/species/c_elegans/gene/WBGene00004735"><em>sbp-1</em></a> in the <a href="http://www.us.lifesciences.sourcebioscience.com/clone-products/non-mammalian/c-elegans/c-elegans-rnai-library/">Ahringer</a> and <a href="http://worfdb.dfci.harvard.edu">ORFeome</a> libraries. This can be done easily with the <code><a href="../reference/rnai.html">rnai()</a></code> function, just search by <strong>name</strong>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/rnai.html">rnai</a></span>(<span class="st">"sbp-1"</span>, <span class="dt">format =</span> <span class="st">"name"</span>) %&gt;%<span class="st"> </span>t</code></pre></div>
<pre><code>##          [,1]                                                                                
## name     "sbp-1"                                                                             
## gene     "WBGene00004735"                                                                    
## sequence "Y47D3B.7"                                                                          
## clone    "ahringer384-III-6-C01, ahringer96-86-B01, orfeome96-10059-H06, orfeome96-11010-G06"</code></pre>
<p>Alternatively, you can search by <a href="http://www.wormbase.org">WormBase</a> <strong>gene</strong> identifier or <strong>sequence</strong>, aka open reading frame (ORF). These queries will return the same result:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/rnai.html">rnai</a></span>(<span class="st">"WBGene00004735"</span>, <span class="dt">format =</span> <span class="st">"gene"</span>)
<span class="kw"><a href="../reference/rnai.html">rnai</a></span>(<span class="st">"Y47D3B.7"</span>, <span class="dt">format =</span> <span class="st">"sequence"</span>)</code></pre></div>
</div>
<div id="mapping-rnai-clones-to-gene-identifiers" class="section level2">
<h2 class="hasAnchor">
<html><body><a href="#mapping-rnai-clones-to-gene-identifiers" class="anchor"> </a></body></html>Mapping RNAi clones to gene identifiers</h2>
<p>The <code><a href="../reference/rnai.html">rnai()</a></code> function was written to simplify the task of identifying wells of interest in an RNAi screen. You can simply input the well identifiers by plate location, and <code><a href="../reference/rnai.html">rnai()</a></code> will return up-to-date gene identifiers from <a href="http://www.wormbase.org">WormBase</a>.</p>
<div id="ahringer-library" class="section level3">
<h3 class="hasAnchor">
<html><body><a href="#ahringer-library" class="anchor"> </a></body></html>
<a href="http://www.us.lifesciences.sourcebioscience.com/clone-products/non-mammalian/c-elegans/c-elegans-rnai-library/">Ahringer</a> library</h3>
<p>Here’s how you map wells from the <a href="http://www.us.lifesciences.sourcebioscience.com/clone-products/non-mammalian/c-elegans/c-elegans-rnai-library/">Ahringer</a> library in 384 well format. Note that plate location identifiers must include the chromosome in roman numeral format at the beginning.</p>
<p>Wells from the <a href="http://www.us.lifesciences.sourcebioscience.com/clone-products/non-mammalian/c-elegans/c-elegans-rnai-library/">Ahringer</a> library in 96 well format are also supported. Since this library has unique plate numbers, the chromosome prefix is not necessary.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">c</span>(<span class="st">"ahringer384-III-6C01"</span>,
  <span class="st">"ahringer96-86-B01"</span>) %&gt;%<span class="st"> </span>rnai %&gt;%<span class="st"> </span>t</code></pre></div>
<pre><code>##          [,1]                   [,2]               
## clone    "ahringer384-III-6C01" "ahringer96-86-B01"
## gene     "WBGene00004735"       "WBGene00004735"   
## sequence "Y47D3B.7"             "Y47D3B.7"         
## name     "sbp-1"                "sbp-1"</code></pre>
</div>
<div id="orfeome-library" class="section level3">
<h3 class="hasAnchor">
<html><body><a href="#orfeome-library" class="anchor"> </a></body></html>
<a href="http://worfdb.dfci.harvard.edu">ORFeome</a> library</h3>
<p><code><a href="../reference/rnai.html">rnai()</a></code> has mapping support for the <a href="http://worfdb.dfci.harvard.edu">ORFeome</a> library, in 96 well format. Plate position should be entered in the same format as seen on the WORFDB. However, the function is flexible, and also supports entries in the historical format (e.g. <code>GHR-11010@G06</code>), or without a separator:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">c</span>(<span class="st">"GHR-11010@G06"</span>,
  <span class="st">"orfeome96-11010-G06"</span>) %&gt;%<span class="st"> </span>rnai %&gt;%<span class="st"> </span>t</code></pre></div>
<pre><code>##          [,1]             [,2]                 
## clone    "GHR-11010@G06"  "orfeome96-11010-G06"
## gene     "WBGene00004735" "WBGene00004735"     
## sequence "Y47D3B.7"       "Y47D3B.7"           
## name     "sbp-1"          "sbp-1"</code></pre>
</div>
</div>
</div>
<div id="create-a-cherrypick-library" class="section level1">
<h1 class="hasAnchor">
<html><body><a href="#create-a-cherrypick-library" class="anchor"> </a></body></html>Create a cherrypick library</h1>
<p>Here’s a how to generate a list of <a href="http://worfdb.dfci.harvard.edu">ORFeome</a> clones that target the unfolded protein response:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/cherrypick.html">cherrypick</a></span>(<span class="st">"unfolded protein response"</span>,
           <span class="dt">ahringer96 =</span> <span class="ot">FALSE</span>,
           <span class="dt">ahringer384 =</span> <span class="ot">FALSE</span>,
           <span class="dt">orfeome96 =</span> <span class="ot">TRUE</span>) %&gt;%<span class="st"> </span>head</code></pre></div>
<pre><code>## # A tibble: 6 × 5
##             gene                   keyword  sequence     name
##            &lt;chr&gt;                     &lt;chr&gt;     &lt;chr&gt;    &lt;chr&gt;
## 1 WBGene00020038 unfolded protein response  R12E2.13     &lt;NA&gt;
## 2 WBGene00002026 unfolded protein response   C12C8.1   hsp-70
## 3 WBGene00004978 unfolded protein response Y47G6A.10    spg-7
## 4 WBGene00007352 unfolded protein response   C06A1.1 cdc-48.1
## 5 WBGene00002147 unfolded protein response   C41C4.4    ire-1
## 6 WBGene00012532 unfolded protein response  Y32F6B.3    crp-1
## # ... with 1 more variables: clone &lt;chr&gt;</code></pre>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li>
<a href="#gene-identifier-queries">Gene identifier queries</a><ul class="nav nav-pills nav-stacked">
<li><a href="#select-options"><code>select</code> options</a></li>
      <li><a href="#gene-query-examples"><code><a href="../reference/gene.html">gene()</a></code> query examples</a></li>
      </ul>
</li>
      <li>
<a href="#rnai-clone-identifier-queries">RNAi clone identifier queries</a><ul class="nav nav-pills nav-stacked">
<li><a href="#finding-clones-by-gene-identifier">Finding clones by gene identifier</a></li>
      <li><a href="#mapping-rnai-clones-to-gene-identifiers">Mapping RNAi clones to gene identifiers</a></li>
      </ul>
</li>
      <li><a href="#create-a-cherrypick-library">Create a cherrypick library</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Michael Steinbaugh.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
