<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>worminfo • worminfo</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/readable/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">worminfo</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/worminfo.html">Get Started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/steinbaugh/worminfo">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>worminfo</h1>
            
            <h4 class="date">29 November 2017</h4>
          </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>Abstract placeholder text.</p>
    </div>
    
<div class="contents">
<div id="overview" class="section level2">
<h2 class="hasAnchor">
<a href="#overview" class="anchor"></a>Overview</h2>
<p>The goal of the <a href="https://github.com/steinbaugh/worminfo">worminfo</a> package is to quickly and simply provide correct, up-to-date annotation data for the entire <em>C. elegans</em> genome. The underlying gene annotations are periodically sourced from <a href="http://www.wormbase.org">WormBase</a>, <a href="http://www.ensembl.org/Caenorhabditis_elegans">Ensembl</a>, and <a href="http://www.pantherdb.org">PANTHER</a>. RNAi clone annotations are sourced from the official <a href="http://worfdb.dfci.harvard.edu">ORFeome</a> and <a href="http://www.us.lifesciences.sourcebioscience.com/clone-products/non-mammalian/c-elegans/c-elegans-rnai-library/">Ahringer</a> repositories. <a href="http://www.wormbase.org">WormBase</a> is updated on a bi-monthly <a href="http://www.wormbase.org/about/release_schedule">release schedule</a>. If the source data appears outdated, file an <a href="https://github.com/steinbaugh/worminfo/issues">update request</a> on our <a href="https://github.com/steinbaugh/worminfo">GitHub repository</a>, and we’ll promptly issue a rebuild. When the package is loaded, the most recent annotation data is downloaded from our server then cached locally for quick access.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">source</span>(<span class="st">"https://bioconductor.org/biocLite.R"</span>)
<span class="kw">biocLite</span>(<span class="kw">c</span>(
    <span class="st">"steinbaugh/basejump"</span>,
    <span class="st">"steinbaugh/worminfo"</span>
))
<span class="kw">biocLite</span>(<span class="st">"tidyverse"</span>)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(basejump)
<span class="kw">library</span>(worminfo)
<span class="kw">library</span>(tidyverse)</code></pre></div>
</div>
<div id="gene-queries" class="section level2">
<h2 class="hasAnchor">
<a href="#gene-queries" class="anchor"></a>Gene queries</h2>
<p><code><a href="../reference/gene.html">gene()</a></code> accepts multiple <code>identifier</code> formats:</p>
<ul>
<li>gene</li>
<li>sequence</li>
<li>name</li>
<li>class</li>
<li>keyword</li>
</ul>
<p>Output from the function is controlled using the <code>select</code> parameter. The function returns a <code>tibble</code>. We advise that subsequent data manipulations be performed with the <a href="http://tidyverse.org">tidyverse</a> collection of packages, <a href="https://cran.r-project.org/package=dplyr">dplyr</a> in particular.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/gene.html">gene</a></span>(<span class="st">"WBGene00004804"</span>, <span class="dt">format =</span> <span class="st">"gene"</span>)
<span class="kw"><a href="../reference/gene.html">gene</a></span>(<span class="st">"T19E7.2"</span>, <span class="dt">format =</span> <span class="st">"sequence"</span>)
<span class="kw"><a href="../reference/gene.html">gene</a></span>(<span class="st">"skn-1"</span>, <span class="dt">format =</span> <span class="st">"name"</span>)</code></pre></div>
<p>These queries all return the same result:</p>
<pre><code>## Observations: 1
## Variables: 3
## $ gene     &lt;chr&gt; "WBGene00004804"
## $ sequence &lt;chr&gt; "T19E7.2"
## $ name     &lt;chr&gt; "skn-1"</code></pre>
<p>You can obtain all genes in a named <code>class</code>. For example, let’s get the current list of <em>daf</em> (<strong>D</strong>auer <strong>A</strong>rrest <strong>F</strong>orming) genes:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/gene.html">gene</a></span>(<span class="st">"daf"</span>, <span class="dt">format =</span> <span class="st">"class"</span>)</code></pre></div>
<pre><code>## # A tibble: 34 x 4
##    class           gene sequence   name
##    &lt;chr&gt;          &lt;chr&gt;    &lt;chr&gt;  &lt;chr&gt;
##  1   daf WBGene00000897  F29C4.1  daf-1
##  2   daf WBGene00000898 Y55D5A.5  daf-2
##  3   daf WBGene00000899  F25E2.5  daf-3
##  4   daf WBGene00000900  C05D2.1  daf-4
##  5   daf WBGene00000901  W01G7.1  daf-5
##  6   daf WBGene00000902  F31F6.5  daf-6
##  7   daf WBGene00000903  B0412.2  daf-7
##  8   daf WBGene00000904 R05D11.1  daf-8
##  9   daf WBGene00000905  T13C5.1  daf-9
## 10   daf WBGene00000906  F23B2.4 daf-10
## # ... with 24 more rows</code></pre>
<p>You can perform a <code>keyword</code> search. For example, let’s get all genes that are involved in the unfolded protein response:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/gene.html">gene</a></span>(<span class="st">"unfolded protein response"</span>, <span class="dt">format =</span> <span class="st">"keyword"</span>)</code></pre></div>
<pre><code>## # A tibble: 65 x 4
##              gene                   keyword  sequence   name
##             &lt;chr&gt;                     &lt;chr&gt;     &lt;chr&gt;  &lt;chr&gt;
##  1 WBGene00000024 unfolded protein response     AC3.3  abu-1
##  2 WBGene00000025 unfolded protein response  F19G12.7  abu-2
##  3 WBGene00000026 unfolded protein response   F31A3.1  abu-3
##  4 WBGene00000027 unfolded protein response   Y5H2A.3  abu-4
##  5 WBGene00000028 unfolded protein response Y105C5A.4  abu-5
##  6 WBGene00000029 unfolded protein response   C03A7.7  abu-6
##  7 WBGene00000030 unfolded protein response   C03A7.8  abu-7
##  8 WBGene00000031 unfolded protein response  C03A7.14  abu-8
##  9 WBGene00000032 unfolded protein response  R09F10.2  abu-9
## 10 WBGene00000033 unfolded protein response   F35A5.3 abu-10
## # ... with 55 more rows</code></pre>
<div id="select-arguments" class="section level3">
<h3 class="hasAnchor">
<a href="#select-arguments" class="anchor"></a><code>select</code> arguments</h3>
<p>The <code><a href="../reference/gene.html">gene()</a></code> function contains annotation data that can be queried using the <code>select</code> parameter:</p>
<ul>
<li>gene</li>
<li>name</li>
<li>sequence</li>
<li>status</li>
<li>otherIDs</li>
<li>descriptionConcise</li>
<li>descriptionProvisional</li>
<li>descriptionDetailed</li>
<li>descriptionAutomated</li>
<li>class</li>
<li>ortholog</li>
<li>rnaiPhenotype</li>
<li>pantherUniprotKB</li>
<li>pantherSubfamily</li>
<li>pantherFamilyName</li>
<li>pantherSubfamilyName</li>
<li>pantherGoMF</li>
<li>pantherGoBP</li>
<li>pantherGoCC</li>
<li>pantherClass</li>
<li>pantherPathway</li>
</ul>
<p>When <code>select</code> is <code>NULL</code>, a simple result is returned:</p>
<ul>
<li>gene</li>
<li>sequence</li>
<li>name</li>
</ul>
<p>Let’s <code>select()</code> the <code>status</code> and <code>otherIDs</code> for <a href="http://www.wormbase.org/species/c_elegans/gene/WBGene00004735"><em>sbp-1</em></a> (bHLH transcription factor):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/gene.html">gene</a></span>(<span class="st">"sbp-1"</span>,
     <span class="dt">format =</span> <span class="st">"name"</span>,
     <span class="dt">select =</span> <span class="kw">c</span>(<span class="st">"status"</span>, <span class="st">"otherIDs"</span>)) <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw"><a href="http://www.rdocumentation.org/packages/tibble/topics/glimpse">glimpse</a></span>()</code></pre></div>
<pre><code>## Observations: 1
## Variables: 5
## $ name     &lt;chr&gt; "sbp-1"
## $ gene     &lt;chr&gt; "WBGene00004735"
## $ sequence &lt;chr&gt; "Y47D3B.7"
## $ status   &lt;chr&gt; "Live"
## $ otherIDs &lt;chr&gt; "Y47D3B.7, sbp-1, lpd-1, hlh-20"</code></pre>
</div>
</div>
<div id="rnai-clone-queries" class="section level2">
<h2 class="hasAnchor">
<a href="#rnai-clone-queries" class="anchor"></a>RNAi clone queries</h2>
<p>RNA interference libraries are commonly used in <em>C. elegans</em> research. RNAi clones are typically referenced by a gene pair or open reading frame (ORF) sequence, which are often difficult to map to an up-to-date gene identifier. We have written the <code><a href="../reference/rnai.html">rnai()</a></code> function to solve this problem. The function enables a researcher to search for RNAi clones by a gene identifier, or map an RNAi clone by plate location and return the up-to-date <a href="http://www.wormbase.org">WormBase</a> gene identifier. The <code><a href="../reference/rnai.html">rnai()</a></code> function has built-in support for the commonly used <a href="http://worfdb.dfci.harvard.edu">ORFeome</a> and <a href="http://www.us.lifesciences.sourcebioscience.com/clone-products/non-mammalian/c-elegans/c-elegans-rnai-library/">Ahringer</a> RNAi libraries.</p>
<div id="find-clones-by-gene-identifier" class="section level3">
<h3 class="hasAnchor">
<a href="#find-clones-by-gene-identifier" class="anchor"></a>Find clones by gene identifier</h3>
<p>Let’s find the locations of <a href="http://www.wormbase.org/species/c_elegans/gene/WBGene00004735"><em>sbp-1</em></a> in the <a href="http://www.us.lifesciences.sourcebioscience.com/clone-products/non-mammalian/c-elegans/c-elegans-rnai-library/">Ahringer</a> and <a href="http://worfdb.dfci.harvard.edu">ORFeome</a> libraries. This can be done easily with the <code><a href="../reference/rnai.html">rnai()</a></code> function, just search by <code>name</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/rnai.html">rnai</a></span>(<span class="st">"sbp-1"</span>, <span class="dt">format =</span> <span class="st">"name"</span>) <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw"><a href="http://www.rdocumentation.org/packages/tibble/topics/glimpse">glimpse</a></span>()</code></pre></div>
<pre><code>## Observations: 1
## Variables: 7
## $ name        &lt;chr&gt; "sbp-1"
## $ gene        &lt;chr&gt; "WBGene00004735"
## $ sequence    &lt;chr&gt; "Y47D3B.7"
## $ genePair    &lt;chr&gt; "Y47D3B.7"
## $ orfeome96   &lt;chr&gt; "10059-H06, 11010-G06"
## $ ahringer384 &lt;chr&gt; "III-6-C01"
## $ ahringer96  &lt;chr&gt; "86-B01"</code></pre>
<p>Alternatively, you can map by <code>gene</code>, <code>sequence</code> (ORF), or legacy <code>genePair</code>. These queries return the same result:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/rnai.html">rnai</a></span>(<span class="st">"WBGene00004735"</span>, <span class="dt">format =</span> <span class="st">"gene"</span>)
<span class="kw"><a href="../reference/rnai.html">rnai</a></span>(<span class="st">"Y47D3B.7"</span>, <span class="dt">format =</span> <span class="st">"sequence"</span>)
<span class="kw"><a href="../reference/rnai.html">rnai</a></span>(<span class="st">"Y47D3B.7"</span>, <span class="dt">format =</span> <span class="st">"genePair"</span>)</code></pre></div>
</div>
<div id="map-clones-by-plate-location" class="section level3">
<h3 class="hasAnchor">
<a href="#map-clones-by-plate-location" class="anchor"></a>Map clones by plate location</h3>
<p>The <code><a href="../reference/rnai.html">rnai()</a></code> function was written to simplify the task of identifying wells of interest in an RNAi screen. You can simply input the well identifiers by plate location, and <code><a href="../reference/rnai.html">rnai()</a></code> will return up-to-date gene identifiers from <a href="http://www.wormbase.org">WormBase</a>.</p>
<div id="ahringer-library-format" class="section level4">
<h4 class="hasAnchor">
<a href="#ahringer-library-format" class="anchor"></a><a href="http://www.us.lifesciences.sourcebioscience.com/clone-products/non-mammalian/c-elegans/c-elegans-rnai-library/">Ahringer</a> library format</h4>
<p>Well identifiers for the <a href="http://www.us.lifesciences.sourcebioscience.com/clone-products/non-mammalian/c-elegans/c-elegans-rnai-library/">Ahringer</a> 384 well library must include a chromosome prefix in roman numeral format. The 96 well <a href="http://www.us.lifesciences.sourcebioscience.com/clone-products/non-mammalian/c-elegans/c-elegans-rnai-library/">Ahringer</a> library has unique plate numbers and does not require a chromosome prefix.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">c</span>(<span class="st">"ahringer384-III-6C01"</span>,
  <span class="st">"ahringer96-86-B01"</span>) <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw"><a href="../reference/rnai.html">rnai</a></span>() <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw"><a href="http://www.rdocumentation.org/packages/tibble/topics/glimpse">glimpse</a></span>()</code></pre></div>
<pre><code>## Observations: 2
## Variables: 4
## $ clone    &lt;chr&gt; "ahringer384-III-6C01", "ahringer96-86-B01"
## $ gene     &lt;chr&gt; "WBGene00004735", "WBGene00004735"
## $ sequence &lt;chr&gt; "Y47D3B.7", "Y47D3B.7"
## $ name     &lt;chr&gt; "sbp-1", "sbp-1"</code></pre>
</div>
<div id="orfeome-library-format" class="section level4">
<h4 class="hasAnchor">
<a href="#orfeome-library-format" class="anchor"></a><a href="http://worfdb.dfci.harvard.edu">ORFeome</a> library format</h4>
<p>We advise formatting identifiers for the 96 well <a href="http://worfdb.dfci.harvard.edu">ORFeome</a> library as <code>orfeome96-10001-A01</code>. However, the function is flexible, and also supports the historical format (e.g. <code>GHR-11010@G06</code>):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">c</span>(<span class="st">"GHR-11010@G06"</span>,
  <span class="st">"orfeome96-11010-G06"</span>) <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw"><a href="../reference/rnai.html">rnai</a></span>() <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw"><a href="http://www.rdocumentation.org/packages/tibble/topics/glimpse">glimpse</a></span>()</code></pre></div>
<pre><code>## Observations: 2
## Variables: 4
## $ clone    &lt;chr&gt; "GHR-11010@G06", "orfeome96-11010-G06"
## $ gene     &lt;chr&gt; "WBGene00004735", "WBGene00004735"
## $ sequence &lt;chr&gt; "Y47D3B.7", "Y47D3B.7"
## $ name     &lt;chr&gt; "sbp-1", "sbp-1"</code></pre>
</div>
</div>
<div id="construct-a-cherrypick-library" class="section level3">
<h3 class="hasAnchor">
<a href="#construct-a-cherrypick-library" class="anchor"></a>Construct a cherrypick library</h3>
<p>Here’s a how to generate a list of <a href="http://worfdb.dfci.harvard.edu">ORFeome</a> clones that target the unfolded protein response:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/cherrypick.html">cherrypick</a></span>(<span class="st">"unfolded protein response"</span>, <span class="dt">format =</span> <span class="st">"keyword"</span>)</code></pre></div>
<pre><code>## # A tibble: 62 x 7
##                      keyword           gene  sequence   name
##                        &lt;chr&gt;          &lt;chr&gt;     &lt;chr&gt;  &lt;chr&gt;
##  1 unfolded protein response WBGene00000024     AC3.3  abu-1
##  2 unfolded protein response WBGene00000025  F19G12.7  abu-2
##  3 unfolded protein response WBGene00000026   F31A3.1  abu-3
##  4 unfolded protein response WBGene00000027   Y5H2A.3  abu-4
##  5 unfolded protein response WBGene00000028 Y105C5A.4  abu-5
##  6 unfolded protein response WBGene00000029   C03A7.7  abu-6
##  7 unfolded protein response WBGene00000030   C03A7.8  abu-7
##  8 unfolded protein response WBGene00000031  C03A7.14  abu-8
##  9 unfolded protein response WBGene00000032  R09F10.2  abu-9
## 10 unfolded protein response WBGene00000033   F35A5.3 abu-10
## # ... with 52 more rows, and 3 more variables: orfeome96 &lt;chr&gt;,
## #   ahringer384 &lt;chr&gt;, ahringer96 &lt;chr&gt;</code></pre>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#overview">Overview</a></li>
      <li><a href="#gene-queries">Gene queries</a></li>
      <li><a href="#rnai-clone-queries">RNAi clone queries</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by <a href="http://steinbaugh.com">Michael Steinbaugh</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
